Draft a Servicing Granularity overview doc

dagood opened the pull request:
This goes into more detail on the problem we're facing with [source-build#923 "Figure out how source-build will work with CoreFX per-package servicing policy"](https://github.com/dotnet/source-build/issues/923).

@dleeapho @dseefeld @crummel @adaggarwal @NikolaMilosavljevic @MichaelSimons

omajid commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,146 @@
+This is the general rule that if a product hasn't changed, it shouldn't be
+rebuilt and rereleased with a new patch version. It's intuitive and ordinary.
+For example, MSBuild doesn't release new NuGet packages based on downstream
+decisions like a new version of the .NET Core SDK being released.
```
Most Linux distributions follow the opposite strategy. Every time a distro package is "released", every component that's part of the source code of the distro package is rebuilt. A component that has source code living in an external distro package is not rebuilt.

Your nuget example is basically:

- Nuget sources are not part of msbuild
- Nuget and msbuild can update/release separately

But the cases we are looking at here **are** part of corefx (or some other repo) sources, but are not "released" from the same branch.

dagood replied:
It's not intended as a NuGet repo example, "NuGet packages" just meant `.nupkg` here. Will clarify:

> For example, MSBuild doesn't release new `.nupkgs` to the NuGet Gallery based on downstream decisions like a new version of the .NET Core SDK being released.

(I don't understand the MSBuild/NuGet interplay interpretation well enough to understand the contrast vs. CoreFX, maybe clarification is enough?)

And yeah... I'll reduce the scope of "it's ordinary" to  NuGet Gallery packages.

dagood replied:
When I look at this more closely:

> Every time a distro package is "released", every component that's part of the source code of the distro package is rebuilt. A component that has source code living in an external distro package is not rebuilt.

This is the same strategy, just with the "product" boundary being the distro package rather than a NuGet package. I'm going to try writing an example in each context to compare and hopefully make this clearer. The difference here is that we're treating the whole .NET Core SDK as one distro package and the Microsoft build is chopped up into many NuGet package "products".

omajid commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,146 @@
+
+> A *product* for the sake of this doc is any released asset where an
+independent decision is made whether that specific asset will be patched and
+shipped.
```
I think this statement points to the root of our troubles: source code is included in a branch but it is not compiled/released with everything else in the branch.

If the "product" was a separate source repository, I think things would be very much more obvious, as is the case in your nuget/msbuild example above.

dagood replied:
@nguerrera mentioned in https://github.com/dotnet/core-sdk/issues/5738#issuecomment-558713055 there's some already talk of actually separating out the targeting pack products at least... so there's more hope to get repo-owner-maintained support for "product subsets" than I thought. I'll update the doc to reflect this.

We can also take different approaches for different products where practical. I mention the CoreFX OOB packages a lot here, and they mess up some approaches, but targeting packs have constraints (only one used per TFM per release) that make them better candidates for this.

dagood replied:
Another approach for targeting packs in particular is generating them in [source-build-reference-packages](https://github.com/dotnet/source-build-reference-packages), where the source code will be static for sure. This creates some maintenance load because when there's a new TFM or patch, the new version will have to be reverse engineered. But it's a known process that the source-build team is already using for the Microsoft.NETCore.App.Ref 3.0.0 package in the 3.0.1 build. (I learned about this today.)

I think it makes sense to go with that for targeting packs until we get a feature in the repos to let us easily go back in time and specifically build the targeting packs.

I'm going to try narrowing this doc down to focus on the CoreFX OOB packages.

omajid commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,146 @@
+
+There is feedback that we need to keep it simple, continue building a ".NET Core
+SDK + Runtime" product. Asking distro maintainers to keep track of many products
+is a significant burden.
```
+1. I dont think it's a good idea to have a complex process with the expectation that everyone gets it right.

omajid commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,146 @@
+
+### Building from source
+This is bad for source-builds, where downloading prebuilt products is forbidden.
+Every upstream product must be built from source.
```
Some (not all) Linux distribution do keep around older distro packages in their distro repository (eg, we build source-build 3.0.10, and 3.0.9 is still around for installation). If that was a general policy (it's not, for example, in Fedora) than source-build could do what Microsoft build does and use older builds. Aside from being not being supported everywhere, it also means source-build needs previous patch versions to have been built in that distro too. A new distro couldn't skip packaging 3.0.0 if 3.0.10 still needs something from there.

dagood replied:
The new distro onboarding is the issue I have with this... written out in the doc below:

> 1. Say the world has changed in a way that makes it impossible to build anything
>    earlier than CoreFX `v3.0.3` on a new distro or platform. Then, patches need
>    to be applied to `v3.0.0`, `v3.0.1`, and `v3.0.2` to get them to build,
>    because the OOB packages are required. The patching work *probably* only
>    needs to be done once and applied on all affected commits, but as the release
>    branch continues to be serviced, the amount of work to get the product built
>    on a new distro balloons.

(This is also not as big of an issue if we only use this approach for targeting packs, not CoreFX OOB packages.)

tmds commented:
imo, we don't want much granularity in source-build.
From maintainer perspective having separate packages adds additional complexity.
There should be a gain/rationale for that complexity.

The repositories that are used by source-build have proper release branches. We should be able to rely on that to have confidence re-built binaries only have minimal changes.

dagood commented:
> The repositories that are used by source-build have proper release branches.

From a Linux distro packaging perspective, they don't, IMO. The way the Microsoft build works, the shape of a repo's outputs on the release branch change over time, depending on what's being patched for the upcoming servicing version Microsoft plans to release. *Forcing* packages to build in the release branch puts us into territory that is unexplored by the Microsoft build and its tests.

We don't have to make extreme decisions based on this, but it's the root of the difference here.

tmds commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,192 @@
+
+This doc is a more detailed description of the problems in [source-build#923
+"Figure out how source-build will work with CoreFX per-package servicing
+policy"](https://github.com/dotnet/source-build/issues/923).
```
From that issue:

> but we risk shipping a different product than the Microsoft build if we just do that:

Because Microsoft is not using source-build, there is a chance we are not shipping the same thing.
That said, we want source-build to be complete in terms of features, quality and security.
Maybe that is doable by building of the proper branches/tags, and doesn't require granular building.

We need anyhow to be able to source-build from scratch.
So granular building requires extra effort in source-build.

dagood replied:
> Maybe that is doable by building of the proper branches/tags, and doesn't require granular building.

It is, but I'm trying to describe in this doc what problems this presents. E.g., source-build probably has to build, say, `v3.0.0`, `v3.0.1`, `v3.0.2`, and `v3.0.3` of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes. (Due to needing to be able to build from scratch, as you mentioned.) Then, each of those tags would need to be maintained to bring on new platforms, fix issues, etc., not something Microsoft has been doing.

Could we change the way the repos build to avoid this? (Have the release branches be "[proper](https://github.com/dotnet/source-build/pull/1389#issuecomment-560933567)"?) Maybe--this will need more investigation and work with each repo, and I'm hoping this doc can help motivate investigation like that to find other more reasonable approaches.

I'll add a section about that... trying to get "proper" release branches is ideal, even though I don't know how to get there and it's counter to a longstanding Microsoft servicing routine. I didn't have a proposal at first, but I guess this would be it.

tmds replied:
> E.g., source-build probably has to build, say, v3.0.0, v3.0.1, v3.0.2, and v3.0.3 of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes. (Due to needing to be able to build from scratch, as you mentioned.)

Maybe we should consider this the driving use-case?

dagood replied:
It is IMO, I have it in the doc as the top reason I think building every dependency node would be unsustainable: 

> 1. Say the world has changed in a way that makes it impossible to build anything   earlier than CoreFX `v3.0.3` on a new distro or platform. Then, patches need   to be applied to `v3.0.0`, `v3.0.1`, and `v3.0.2` to get them to build,   because the OOB packages are required. The patching work *probably* only   needs to be done once and applied on all affected commits, but as the release   branch continues to be serviced, the amount of work to get the product built   on a new distro balloons.

Did you have somewhere else in mind this drives a specific direction?

tmds replied:
Sorry for the slow response rate. I've been too-much-multi-tasking.

>  E.g., source-build probably has to build, say, v3.0.0, v3.0.1, v3.0.2, and v3.0.3 of CoreFX in order to get a prebuiltless SDK with working 3.0.3 runtimes

I mean: we should try to make it possible to build .NET Core without having to do builds of multiple versions of a single repository.

I think we are looking at this from the way Microsoft is creating packages for Microsoft build (https://github.com/dotnet/source-build/issues/923).

Maybe we should reverse perspective and look from source-build side:
- What different source packages make sense?
- How can these source packages be assembled from the separate repositories?

> The current workaround: assume tip source code can build old packages

Maybe we can do this without having to call it a workaround?

dagood replied:
No worries at all, been very busy myself with dotnet/runtime consolidation.

> Maybe we should reverse perspective and look from source-build side:
> 
> * What different source packages make sense?
> * How can these source packages be assembled from the separate repositories?

I think it would be interesting to see if a different arrangement of the package granularity would make more sense for source-build, and ideally have the Microsoft build adopt it. I'm a little skeptical that will provide enough additional value over the existing Microsoft build granularity, though. (In the context of needing value to justify the overhead for .NET Core distro maintainers having to track down and piece together multiple source packages.)

> > The current workaround: assume tip source code can build old packages
> 
> Maybe we can do this without having to call it a workaround?

I'm ok with that if we can get rid of "assume", which to me seems like a pretty big deal. (This is what I believe having "proper" release branches would mean.)

nguerrera replied:
In an ideal world anything that ships on its own schedule would have its own repo/branch. There are many reasons why we don't live in that world and probably never will. Unfortunately. On the other end,  re-shipping absolutely everything every time anything in any of the repos changes has real problems too.

I think maybe we can improve "assume" to "validate". In any given month any given package might need to ship, so the code has to be sitting ready to take a  patch. Thus the code in the branch should be trusted as good code matching the last time it shipped until a change is made to ship it again. If a source build distributor would pref to rebuild that code every month, they can do that if they trust that it is good, which we all should based on above reasoning. That does not mean that Microsoft should have to choose to ship the same payload every month to nuget.org for packages that haven't changed.

It terms of validation, I could imagine running tests with the LKG version of a package and the live version always, but not pushing the live version to transport feeds for MS builds or nuget.org, unless there's an explicit opt in. I could further imagine analyzing if source/API has changed, but the "ship it this month" bit isn't flipped, etc.

This is complicated, for sure. I just think that it is not easy to land on either of the simpler solutions: ship the world everywhere always, or separate schedule == separate repo.

tmds commented on Documentation/ServicingGranularity.md:
```diff
@@ -0,0 +1,192 @@
+Linux distributions tend to follow the same strategy, where the product is a
+distro package. To create a distro package release, all the source code making
+it up is rebuilt. A component that has source code living in an external distro
+package is not rebuilt.
```
A single source package is providing the sources for multiple 'distro' packages.
This is also what we are doing with .NET Core.

It's not clear for me how this should work with granular builds.

dagood replied:
Is there a good name for the source package + all distro packages that come from it? This is what I'm referring to here--each new source package produces a set of distro packages that is "a product". All of the distro packages are built and released at the same time, effectively.

If source-build were granular, I believe we'd have many smaller source packages. (And fewer (maybe just one) distro package per source package.)

tmds replied:
I don't know the correct name, maybe @omajid has a suggestion.
I'm fine with calling it a product.

Earlier in the text it says:

> Linux distributions tend to follow the same strategy, where the product is a distro package.

This should then be:

> Linux distributions tend to follow the same strategy, where the product are all the packages built from a single source package.

dagood commented:
I reworked it to focus on the recommended approach and moved the bulk of the doc (explaining why granularity is bad for source-build) into a different file. It should be much more readable now.

dagood commented:
I moved the doc to fit in better with `Documentation/planning/arcade-powered-source-build/README.md` and added a small reference. This is something we should consider for 5.0--it would be bad to hit one of these pitfalls as we're getting the source-built SDK into more distros. I believe this should be incremental work after getting the initial source-build-in-arcade plan done. Once we merge the doc we can write up issues that reference it as appropriate.

@dseefeld PTAL
